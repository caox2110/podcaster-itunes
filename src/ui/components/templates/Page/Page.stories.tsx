import type { ComponentStory, ComponentMeta } from '@storybook/react';
import { withPerformance } from 'storybook-addon-performance';

import Page from '.';

const Template: ComponentStory<typeof Page> = (properties) => (
  <Page {...properties}>
    <h1>Testing Page</h1>
  </Page>
);

const PageWithTitle = Template.bind({});
PageWithTitle.args = {
  title: 'Main Page',
};

export { PageWithTitle };
export default {
  title: 'UI/Templates/Page',
  component: Page,
  decorators: [withPerformance],
} as ComponentMeta<typeof Page>;
