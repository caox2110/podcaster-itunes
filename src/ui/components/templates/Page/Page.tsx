import { forwardRef, Ref, ReactNode, useMemo } from 'react';
import { Helmet } from 'react-helmet-async';

import { constants } from '../../../../helper';
import { MuiBox } from '../..';

/**
 * Page component props definition
 *
 * @interface PageProperties
 * @typedef {PageProperties}
 */
interface PageProperties {
  /**
   * Component to render
   *
   * @type {ReactNode}
   */
  children: ReactNode;
  /**
   * Title for page
   *
   * @type {string}
   */
  title: string;
  /**
   * Any others meta tag to add
   *
   * @type {?ReactNode}
   */
  meta?: ReactNode;
}

const {
  app: { APP_NAME },
} = constants;

/**
 * Page template that helmet to update the title of the page after navigation.
 *
 * @param {PageProperties} { children, title = '', meta, ...rest }
 * @param {Ref<unknown>} reference
 * @returns {*}
 */
function Page(
  { children, title = '', meta = null, ...rest }: PageProperties,
  reference: Ref<unknown>,
) {
  const fullTitle = useMemo<string>(() => `${title} | ${APP_NAME}`, [title]);

  return (
    <>
      <Helmet defaultTitle={APP_NAME}>
        <title>{fullTitle}</title>
        {meta}
      </Helmet>
      <MuiBox ref={reference} {...rest}>
        {children}
      </MuiBox>
    </>
  );
}

// export type { PageProperties };
export default forwardRef(Page);
