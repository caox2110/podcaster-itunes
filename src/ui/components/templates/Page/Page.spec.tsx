import { render, waitFor } from '@testing-library/react';
import type { RenderResult } from '@testing-library/react';

import { ThemeContainer } from '../..';

import Page from '.';

function renderApp(): RenderResult {
  const utils = render(
    <ThemeContainer>
      <Page title='Main Page'>
        <h1>Testing text</h1>
      </Page>
    </ThemeContainer>,
  );
  return {
    ...utils,
  };
}

describe('Page template rendering', () => {
  it('Should render without throwing any errors', () => {
    const { baseElement, container } = renderApp();

    expect(baseElement).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
  it('Should load a title after the page rendering', async () => {
    renderApp();
    await waitFor(() => expect(document.title).toEqual('Main Page | Podcaster'));
  });
});
