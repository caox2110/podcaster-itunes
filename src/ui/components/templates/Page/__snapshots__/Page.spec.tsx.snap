// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page template rendering Should render without throwing any errors 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <h1>
      Testing text
    </h1>
  </div>
</div>
`;
